In [1]:

import pandas as pd
import numpy as np

print("pandas version:", pd.__version__)
print("numpy version:", np.__version__)
print("Libraries loaded successfully!")

pandas version: 2.3.3
numpy version: 2.3.5
Libraries loaded successfully!


In [22]:
# Load the dataset
df = pd.read_excel(r'C:\Users\User\Desktop\DataAnalysesProject\data\raw\Online Retail.xlsx')

# First look
print("Shape of dataset:", df.shape)
print("Columns:", df.columns.tolist())

Shape of dataset: (541909, 8)
Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


In [17]:
import os
print(os.getcwd())

C:\Users\User\Desktop\DataAnalysesProject\notebook


In [20]:
import os
# This will show all files on your Desktop
desktop = os.path.join(os.path.expanduser("~"), "Desktop")
print(desktop)
print(os.listdir(desktop))

C:\Users\User\Desktop
['#Stereo_ Left and Right Stereo Sound Test - Copy.mp4', '#Stereo_ Left and Right Stereo Sound Test.mp4', '.ipynb_checkpoints', 'Brave.lnk', 'cpp assingment.docx', 'CYBER SECURITY', 'DataAnalysesProject', 'DataAnalystProject', 'DBMS project.txt', 'desktop.ini', 'Eclipse IDE for Enterprise Java and Web Developers - 2025-06.lnk', 'HTML Tutorial', 'Keyboard Keys Checker.exe', 'labpython', 'Microsoft Edge.lnk', 'NewPhyton', 'passportAutomationSystem', 'pavani', 'phypro', 'phyton', 'practice', 'PYTHON 1', 'python-first-project', 'rakshitha', 'Rakshitha - Chrome.lnk', 'Scratch 3.lnk', 'SE LAB PRO.docx', 'SOFTWARE', 'superstore-dataset', 'udemy', 'Untitled.ipynb', 'Untitled1.ipynb', 'Visual Studio Code.lnk', 'webtecnology', 'WOMENSTUDIES.docx', '~$ LAB PRO.docx', '~$MENSTUDIES.docx', '~$p assingment.docx']


In [23]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [25]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [26]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386048,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


In [27]:
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

In [29]:
df['Country'].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


# E-Commerce Retail Sales Analytics
## Phase 2 - Data Loading & Exploration
### Dataset: UCI Online Retail Dataset
**Date:** 10-June-2026

**What this notebook covers:**
- Loading the dataset
- Checking shape, columns, data types
- Identifying null values
- Basic statistical summary
- 

## Day 4 - Data Cleaning
**Date:** 11-June-2026


In [1]:
# Import libraries
import pandas as pd
import numpy as np

# Load dataset
df = pd.read_excel(r'C:\Users\User\Desktop\DataAnalysesProject\data\raw\Online Retail.xlsx')

print("Dataset loaded!")
print("Shape:", df.shape)

Dataset loaded!
Shape: (541909, 8)


In [2]:
# Check how many cancelled orders exist BEFORE cleaning
print("Before removing cancellations:", df.shape)
print("Cancelled orders (Quantity < 0):", (df['Quantity'] < 0).sum())

# Remove rows where Quantity is less than 0
df = df[df['Quantity'] > 0]

print("After removing cancellations:", df.shape)

Before removing cancellations: (541909, 8)
Cancelled orders (Quantity < 0): 10624
After removing cancellations: (531285, 8)


In [3]:
# Check how many zero price rows exist
print("Zero UnitPrice rows:", (df['UnitPrice'] <= 0).sum())

# Remove rows where UnitPrice is 0 or negative
df = df[df['UnitPrice'] > 0]

print("After removing zero prices:", df.shape)

Zero UnitPrice rows: 1181
After removing zero prices: (530104, 8)


In [4]:
# Check missing CustomerIDs
print("Missing CustomerIDs:", df['CustomerID'].isnull().sum())

# Remove rows where CustomerID is blank
df = df.dropna(subset=['CustomerID'])

print("After removing missing CustomerIDs:", df.shape)

Missing CustomerIDs: 132220
After removing missing CustomerIDs: (397884, 8)


In [5]:
# List of junk descriptions to remove
junk = ['AMAZON FEE', 'Manual', 'SAMPLES', 'Damaged', 
        'Bank Charges', 'POSTAGE', 'DOTCOM POSTAGE',
        'Adjust bad debt', 'CRUK Commission']

# Check how many junk rows exist
print("Junk rows:", df['Description'].isin(junk).sum())

# Remove junk rows
df = df[~df['Description'].isin(junk)]

print("After removing junk:", df.shape)

Junk rows: 1411
After removing junk: (396473, 8)


In [6]:
# Fix InvoiceDate from text to datetime
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

# Fix CustomerID from float to integer
df['CustomerID'] = df['CustomerID'].astype(int)

# Verify the fix
print("Data types after fixing:")
print(df[['InvoiceDate', 'CustomerID']].dtypes)

Data types after fixing:
InvoiceDate    datetime64[ns]
CustomerID              int64
dtype: object


In [7]:
# Create TotalRevenue column
df['TotalRevenue'] = df['Quantity'] * df['UnitPrice']

# Extract Month from InvoiceDate
# dt.month gives month number (1=Jan, 2=Feb etc)
df['Month'] = df['InvoiceDate'].dt.month

# Extract Year from InvoiceDate
df['Year'] = df['InvoiceDate'].dt.year

# Extract DayOfWeek from InvoiceDate
# 0=Monday, 1=Tuesday ... 6=Sunday
df['DayOfWeek'] = df['InvoiceDate'].dt.dayofweek

# Verify new columns
print("New columns added!")
print(df[['Quantity', 'UnitPrice', 'TotalRevenue', 
          'Month', 'Year', 'DayOfWeek']].head())

New columns added!
   Quantity  UnitPrice  TotalRevenue  Month  Year  DayOfWeek
0         6       2.55         15.30     12  2010          2
1         6       3.39         20.34     12  2010          2
2         8       2.75         22.00     12  2010          2
3         6       3.39         20.34     12  2010          2
4         6       3.39         20.34     12  2010          2


In [8]:
# 1. Check outliers in Quantity
print("Quantity stats:")
print("Max:", df['Quantity'].max())
print("Min:", df['Quantity'].min())
print("Mean:", round(df['Quantity'].mean(), 2))

# 2. np.percentile - finds the value at a given percentage
# 99th percentile means - 99% of values are below this number
q99 = np.percentile(df['Quantity'], 99)
print("\n99th percentile of Quantity:", q99)

# 3. np.clip - caps values between a minimum and maximum
# Any Quantity above 99th percentile gets capped
df['Quantity_clipped'] = np.clip(df['Quantity'], 1, q99)
print("Quantity clipped successfully!")

# 4. np.log - log transform of TotalRevenue
# Useful for visualization when data is skewed
df['LogRevenue'] = np.log(df['TotalRevenue'] + 1)
print("Log Revenue created!")

# Verify
print("\nSample:")
print(df[['Quantity', 'Quantity_clipped', 
          'TotalRevenue', 'LogRevenue']].head())

Quantity stats:
Max: 80995
Min: 1
Mean: 13.01

99th percentile of Quantity: 120.0
Quantity clipped successfully!
Log Revenue created!

Sample:
   Quantity  Quantity_clipped  TotalRevenue  LogRevenue
0         6                 6         15.30    2.791165
1         6                 6         20.34    3.060583
2         8                 8         22.00    3.135494
3         6                 6         20.34    3.060583
4         6                 6         20.34    3.060583


In [9]:
# Define output path
output_path = r'C:\Users\User\Desktop\DataAnalysesProject\data\clean\online_retail_clean.csv'

# Save cleaned dataframe to CSV
df.to_csv(output_path, index=False)

print("✅ Clean data saved successfully!")
print("Location:", output_path)
print("Final shape:", df.shape)
print("Columns:", df.columns.tolist())

✅ Clean data saved successfully!
Location: C:\Users\User\Desktop\DataAnalysesProject\data\clean\online_retail_clean.csv
Final shape: (396473, 14)
Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country', 'TotalRevenue', 'Month', 'Year', 'DayOfWeek', 'Quantity_clipped', 'LogRevenue']


## Day 4 Cleaning Summary
- Started with: 541,909 rows
- Removed cancelled orders: 10,624 rows
- Removed zero UnitPrice: 1,181 rows
- Removed missing CustomerIDs: 132,220 rows
- Removed junk descriptions: 1,411 rows
- Final clean dataset: 396,473 rows
- New columns added: TotalRevenue, Month, Year, 
  DayOfWeek, Quantity_clipped, LogRevenue
- Clean file saved to: data/clean/online_retail_clean.csv
- 

## Day 5 - RFM Analysis
**Date:** 11-June-2026

In [10]:
import pandas as pd
import numpy as np

# Load CLEAN data today - not raw!
df = pd.read_csv(r'C:\Users\User\Desktop\DataAnalysesProject\data\clean\online_retail_clean.csv')

# Fix InvoiceDate back to datetime after loading CSV
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

print("Clean data loaded!")
print("Shape:", df.shape)
print("Date range:", df['InvoiceDate'].min(), "to", df['InvoiceDate'].max())

Clean data loaded!
Shape: (396473, 14)
Date range: 2010-12-01 08:26:00 to 2011-12-09 12:50:00


In [11]:
# Reference date = last date in dataset + 1 day
reference_date = df['InvoiceDate'].max() + pd.Timedelta(days=1)
print("Reference date:", reference_date)

# Calculate RFM for each customer
rfm = df.groupby('CustomerID').agg(
    Recency   = ('InvoiceDate', lambda x: (reference_date - x.max()).days),
    Frequency = ('InvoiceNo', 'nunique'),
    Monetary  = ('TotalRevenue', 'sum')
).reset_index()

print("\nRFM table created!")
print("Shape:", rfm.shape)
print("\nFirst 5 customers:")
print(rfm.head())

Reference date: 2011-12-10 12:50:00

RFM table created!
Shape: (4334, 4)

First 5 customers:
   CustomerID  Recency  Frequency  Monetary
0       12346      326          1  77183.60
1       12347        2          7   4310.00
2       12348       75          4   1437.24
3       12349       19          1   1457.55
4       12350      310          1    294.40


In [12]:
# Calculate percentile boundaries using NumPy
r_percentiles = np.percentile(rfm['Recency'],   [25, 50, 75])
f_percentiles = np.percentile(rfm['Frequency'], [25, 50, 75])
m_percentiles = np.percentile(rfm['Monetary'],  [25, 50, 75])

print("Recency boundaries:",   r_percentiles)
print("Frequency boundaries:", f_percentiles)
print("Monetary boundaries:",  m_percentiles)

# Score Recency - REVERSED (low recency = better score)
rfm['R_Score'] = np.digitize(rfm['Recency'],
                 bins=r_percentiles, right=True)
rfm['R_Score'] = 4 - rfm['R_Score']  # Reverse the score

# Score Frequency - normal (high frequency = better score)
rfm['F_Score'] = np.digitize(rfm['Frequency'],
                 bins=f_percentiles, right=True)

# Score Monetary - normal (high monetary = better score)  
rfm['M_Score'] = np.digitize(rfm['Monetary'],
                 bins=m_percentiles, right=True)

print("\nRFM Scores added!")
print(rfm.head())

Recency boundaries: [ 18.  51. 143.]
Frequency boundaries: [1. 2. 5.]
Monetary boundaries: [ 305.8675  668.125  1631.6225]

RFM Scores added!
   CustomerID  Recency  Frequency  Monetary  R_Score  F_Score  M_Score
0       12346      326          1  77183.60        1        0        3
1       12347        2          7   4310.00        4        3        3
2       12348       75          4   1437.24        2        2        2
3       12349       19          1   1457.55        3        0        2
4       12350      310          1    294.40        1        0        0


In [13]:
# Create combined RFM score
rfm['RFM_Score'] = rfm['R_Score'] + rfm['F_Score'] + rfm['M_Score']

# Create segments based on scores
def assign_segment(row):
    if row['R_Score'] >= 3 and row['F_Score'] >= 2:
        return 'Recent Active'
    elif row['F_Score'] >= 3 and row['M_Score'] >= 2:
        return 'Frequent Buyers'
    elif row['M_Score'] >= 3:
        return 'High Value'
    elif row['R_Score'] <= 1 and row['F_Score'] <= 1:
        return 'Lost Customers'
    else:
        return 'Average Customers'

# Apply segment function to each row
rfm['Segment'] = rfm.apply(assign_segment, axis=1)

# Count customers in each segment
print("Customer Segments:")
print(rfm['Segment'].value_counts())
print("\nTotal customers:", len(rfm))

Customer Segments:
Segment
Average Customers    1677
Recent Active        1443
Lost Customers        924
High Value            183
Frequent Buyers       107
Name: count, dtype: int64

Total customers: 4334


In [14]:
# Save RFM table to clean folder
rfm_path = r'C:\Users\User\Desktop\DataAnalysesProject\data\clean\rfm_segments.csv'

rfm.to_csv(rfm_path, index=False)

print("✅ RFM data saved!")
print("Columns:", rfm.columns.tolist())
print("\nFinal RFM table sample:")
print(rfm.head(10))

✅ RFM data saved!
Columns: ['CustomerID', 'Recency', 'Frequency', 'Monetary', 'R_Score', 'F_Score', 'M_Score', 'RFM_Score', 'Segment']

Final RFM table sample:
   CustomerID  Recency  Frequency  Monetary  R_Score  F_Score  M_Score  \
0       12346      326          1  77183.60        1        0        3   
1       12347        2          7   4310.00        4        3        3   
2       12348       75          4   1437.24        2        2        2   
3       12349       19          1   1457.55        3        0        2   
4       12350      310          1    294.40        1        0        0   
5       12352       36          7   1385.74        3        3        2   
6       12353      204          1     89.00        1        0        0   
7       12354      232          1   1079.40        1        0        2   
8       12355      214          1    459.40        1        0        1   
9       12356       23          3   2487.43        3        2        3   

   RFM_Score            S

In [15]:
# Statistical summary of RFM
print("=" * 50)
print("RFM STATISTICAL SUMMARY")
print("=" * 50)

print("\n📊 RECENCY (days since last purchase):")
print(f"  Best customer:  {rfm['Recency'].min()} days ago")
print(f"  Worst customer: {rfm['Recency'].max()} days ago")
print(f"  Average:        {round(rfm['Recency'].mean(), 1)} days")
print(f"  Median:         {rfm['Recency'].median()} days")

print("\n📊 FREQUENCY (number of orders):")
print(f"  Most orders:    {rfm['Frequency'].max()}")
print(f"  Least orders:   {rfm['Frequency'].min()}")
print(f"  Average orders: {round(rfm['Frequency'].mean(), 1)}")
print(f"  Median orders:  {rfm['Frequency'].median()}")

print("\n📊 MONETARY (total spent):")
print(f"  Highest spender: £{round(rfm['Monetary'].max(), 2)}")
print(f"  Lowest spender:  £{round(rfm['Monetary'].min(), 2)}")
print(f"  Average spend:   £{round(rfm['Monetary'].mean(), 2)}")
print(f"  Median spend:    £{round(rfm['Monetary'].median(), 2)}")

print("\n📊 SEGMENT SUMMARY:")
segment_summary = rfm.groupby('Segment').agg(
    Count    = ('CustomerID', 'count'),
    Avg_Recency   = ('Recency', 'mean'),
    Avg_Frequency = ('Frequency', 'mean'),
    Avg_Monetary  = ('Monetary', 'mean')
).round(1)
print(segment_summary)

RFM STATISTICAL SUMMARY

📊 RECENCY (days since last purchase):
  Best customer:  1 days ago
  Worst customer: 374 days ago
  Average:        92.7 days
  Median:         51.0 days

📊 FREQUENCY (number of orders):
  Most orders:    206
  Least orders:   1
  Average orders: 4.2
  Median orders:  2.0

📊 MONETARY (total spent):
  Highest spender: £279138.02
  Lowest spender:  £3.75
  Average spend:   £2023.02
  Median spend:    £668.12

📊 SEGMENT SUMMARY:
                   Count  Avg_Recency  Avg_Frequency  Avg_Monetary
Segment                                                           
Average Customers   1677         66.7            1.9         523.7
Frequent Buyers      107         98.1            8.0        2874.0
High Value           183        106.7            3.0        4421.9
Lost Customers       924        253.9            1.2         348.6
Recent Active       1443         17.5            8.8        4470.4


## Day 5 - RFM Analysis Summary

**Total customers analyzed:** 4,334
**Reference date:** 2011-12-10

### Key Findings:
- Best customer bought 1 day ago
- Top customer made 206 orders in one year
- Highest spender: £279,138
- Recent Active = best segment (£4,470 avg spend)
- 924 Lost Customers need win-back campaign
- RFM saved to: data/clean/rfm_segments.csv

## Day 6 - Final Data Preparation & Export
**Date:** 11-June-2026

In [17]:
import pandas as pd
import numpy as np

# Load clean data
df = pd.read_csv(r'C:\Users\User\Desktop\DataAnalysesProject\data\clean\online_retail_clean.csv')

# Fix datetime
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

print("Clean data loaded!")
print("Shape:", df.shape)

Clean data loaded!
Shape: (396473, 14)


In [19]:
# Add Month Name column
# %B gives full month name (January, February etc)
# %b gives short month name (Jan, Feb etc)
df['MonthName'] = df['InvoiceDate'].dt.strftime('%b')

# Add Month-Year column (useful for time series charts)
df['MonthYear'] = df['InvoiceDate'].dt.strftime('%b-%Y')

print("Month columns added!")
print(df[['InvoiceDate', 'Month', 'MonthName', 'MonthYear']].head(10))

Month columns added!
          InvoiceDate  Month MonthName MonthYear
0 2010-12-01 08:26:00     12       Dec  Dec-2010
1 2010-12-01 08:26:00     12       Dec  Dec-2010
2 2010-12-01 08:26:00     12       Dec  Dec-2010
3 2010-12-01 08:26:00     12       Dec  Dec-2010
4 2010-12-01 08:26:00     12       Dec  Dec-2010
5 2010-12-01 08:26:00     12       Dec  Dec-2010
6 2010-12-01 08:26:00     12       Dec  Dec-2010
7 2010-12-01 08:28:00     12       Dec  Dec-2010
8 2010-12-01 08:28:00     12       Dec  Dec-2010
9 2010-12-01 08:34:00     12       Dec  Dec-2010


In [20]:
# Add IsWeekend column using numpy where
# DayOfWeek 5=Saturday, 6=Sunday
df['IsWeekend'] = np.where(df['DayOfWeek'] >= 5, True, False)

# Check distribution
print("Weekend vs Weekday orders:")
print(df['IsWeekend'].value_counts())
print("\nPercentage:")
print(round(df['IsWeekend'].value_counts(normalize=True) * 100, 1))

Weekend vs Weekday orders:
IsWeekend
False    333817
True      62656
Name: count, dtype: int64

Percentage:
IsWeekend
False    84.2
True     15.8
Name: proportion, dtype: float64


In [21]:
# Extract hour from InvoiceDate
df['Hour'] = df['InvoiceDate'].dt.hour

# Check peak hours
print("Orders by Hour:")
print(df['Hour'].value_counts().sort_index())

Orders by Hour:
Hour
6         1
7       376
8      8636
9     21818
10    37799
11    48942
12    71863
13    63825
14    53936
15    45201
16    24007
17    13031
18     2919
19     3318
20      801
Name: count, dtype: int64


In [22]:
print("=" * 50)
print("FINAL DATASET VERIFICATION")
print("=" * 50)

print("\n✅ Shape:", df.shape)
print("\n✅ Columns:", df.columns.tolist())
print("\n✅ Data types:")
print(df.dtypes)
print("\n✅ Null values:")
print(df.isnull().sum())
print("\n✅ Sample data:")
print(df.head(3))

FINAL DATASET VERIFICATION

✅ Shape: (396473, 18)

✅ Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country', 'TotalRevenue', 'Month', 'Year', 'DayOfWeek', 'Quantity_clipped', 'LogRevenue', 'MonthName', 'MonthYear', 'IsWeekend', 'Hour']

✅ Data types:
InvoiceNo                    int64
StockCode                   object
Description                 object
Quantity                     int64
InvoiceDate         datetime64[ns]
UnitPrice                  float64
CustomerID                   int64
Country                     object
TotalRevenue               float64
Month                        int64
Year                         int64
DayOfWeek                    int64
Quantity_clipped             int64
LogRevenue                 float64
MonthName                   object
MonthYear                   object
IsWeekend                     bool
Hour                         int32
dtype: object

✅ Null values:
InvoiceNo           0
StockCo

# Save final clean dataset
final_path = r'C:\Users\User\Desktop\DataAnalysesProject\data\clean\online_retail_final.csv'

df.to_csv(final_path, index=False)

print("✅ Final clean dataset saved!")
print("📁 Location:", final_path)
print("📊 Shape:", df.shape)
print("\n🎯 Phase 2 - Python Cleaning COMPLETE!")
print("=" * 50)
print("Summary:")
print(f"  Raw data rows:   541,909")
print(f"  Clean data rows: {len(df):,}")
print(f"  Rows removed:    {541909 - len(df):,}")
print(f"  Columns added:   10 new columns")
print(f"  Nulls remaining: 0")

## Phase 2 Complete! ✅
### Files created:
- online_retail_clean.csv (basic cleaning)
- rfm_segments.csv (RFM analysis)  
- online_retail_final.csv (final dataset)

### Key achievements:
- Removed 145,436 dirty rows (26% of data)
- Fixed all data types
- Created 10 new useful columns
- Identified 5 customer segments
- Zero nulls in final dataset